### ISSUE TITLE
"genimage ubuntu18.04.1-x86_64-netboot-compute" generates broken image

### ISSUE THREAD

**Issue by jamesfreeman959** [xcat2/xcat-core#5887]
Hi all,

I have been working with xcat 2.14.5 on Ubuntu Server 18.04.1 LTS for a few days and have completed most of the steps required to network boot a SuperMicro motherboard using a stateless image.

As part of the build process, I have copied the image from the loopback mounted ISO:

```# /opt/xcat/bin/copycds-cdrom /iso/ubuntu-18.04.1-server-amd64.iso```

And then generated the netboot image:

```# genimage ubuntu18.04.1-x86_64-netboot-compute```

The process appears to start well and the output from the command looks fine, but right at the end of the process I see the following line repeated over and over:

```ldd: ./linux-vdso.so.1: No such file or directory```

Just before this happens, the output from the command is:

```
Processing triggers for libc-bin (2.27-3ubuntu1) ...
Umount /proc, /dev, /sys, pkgdir and otherpkgdir to the rootimg.
Setting up the locales
Generating locales (this might take a while)...
  en_US.ISO-8859-1... done
  en_US.UTF-8... done
Generation complete.
Generating locales (this might take a while)...
  en_US.ISO-8859-1... done
  en_US.UTF-8... done
Generation complete.
Added ptp.ko as an autodetected dependency
Added pps_core.ko as an autodetected dependency
Added dca.ko as an autodetected dependency
Added i2c-algo-bit.ko as an autodetected dependency
Added mdio.ko as an autodetected dependency
Added libcrc32c.ko as an autodetected dependency
ldd: ./linux-vdso.so.1: No such file or directory
ldd: ./linux-vdso.so.1: No such file or directory
```

The image then seems to complete:

```
ldd: ./linux-vdso.so.1: No such file or directory
cd /tmp/xcatinitrd.23297;find .|cpio -H newc -o|gzip -9 -c - > /install/netboot/ubuntu18.04.1/x86_64/compute/initrd-stateless.gz
685063 blocks
```

And I then pack the image:

```
# packimage ubuntu18.04.1-x86_64-netboot-compute
Packing contents of /install/netboot/ubuntu18.04.1/x86_64/compute/rootimg
archive method:cpio
compress method:gzip
```

Apart from the linux-vdso.so.1 errors, everything looks like it have worked successfully. I then configure my node for network boot and power it on:

```
# rsetboot com1 net -u 
# nodeset com1 osimage=ubuntu18.04.1-x86_64-netboot-compute 
# rpower com1 boot
```

The node powers on, and loads the UEFI network stack, which in turn obtains an IP address, loads the boot image and boots the kernel. 13 seconds into the boot process, everything simply hangs and the boot does not proceed any further. The last drivers loaded by the kernel were the network drivers - however I am not clear whether the hang is driver related, or because it was trying to move on from the kernel to the next phase of the boot. I am a little concerned about the linux-vdso.so.1 errors - I am not sure if these mean the stateless boot image has an error in it, or whether it these can safely be ignored. The presence of the errors, combined with a boot image that doesn't quite get there does make me think something went wrong during the genimage phase. Can anyone confirm if this is indeed the case, and whether this is an issue, or simply something I need to deal with in a different manner?

Thanks in advance for your help,

James

**Comment by cxhong** [comment]
for ppc64el,  we need to download mini.iso.  I think it apply to x86_64 images also.

try to download ubuntu mini.iso.
http://archive.ubuntu.com/ubuntu/dists/bionic-updates/main/installer-amd64/current/images/netboot/

and follow this instruction for ubuntu image:
https://xcat-docs.readthedocs.io/en/stable/guides/admin-guides/manage_clusters/ppc64le/diskless/copy_image.html?highlight=mini.iso

**Comment by bybai** [comment]
@jamesfreeman959,
1.  Confirm you answer:
For x86_64, we do not need to download `mini.iso`. 
I can reproduce genimage error as " ldd: ./linux-vdso.so.1: No such file or directory" on ubuntu18.04.1 x86_64 management node. But I think it is not a block issue for genimage or diskless provision, you can ignore the message. My diskless compute node can be provisioned successfully using commands.
```
/# packimage ubuntu18.04.1-x86_64-netboot-compute
Packing contents of /install/netboot/ubuntu18.04.1/x86_64/compute/rootimg
archive method:cpio
compress method:pigz
/# rinstall c910f04x35v04 osimage=ubuntu18.04.1-x86_64-netboot-compute
```
Since I cannot find `linux-vdso.so.1` from the following ubuntu18.04.1 repo,  genimage print cannot find `linux-vdso.so.1`, 
```
/# cat /etc/apt/sources.list
deb http://archive.ubuntu.com/ubuntu bionic main restricted universe
deb http://archive.ubuntu.com/ubuntu bionic-updates main restricted universe
deb http://archive.ubuntu.com/ubuntu bionic-security main restricted universe
/# apt-get update
Get:1 file:/install/xcat/xcat-core bionic InRelease [2568 B]
Get:2 file:/install/xcat/xcat-dep bionic InRelease [2561 B]
Get:1 file:/install/xcat/xcat-core bionic InRelease [2568 B]
Get:2 file:/install/xcat/xcat-dep bionic InRelease [2561 B]
Hit:3 http://archive.ubuntu.com/ubuntu bionic InRelease
Get:4 http://archive.ubuntu.com/ubuntu bionic-updates InRelease [88.7 kB]
Get:5 http://archive.ubuntu.com/ubuntu bionic-security InRelease [83.2 kB]
Fetched 172 kB in 2s (93.4 kB/s)
Reading package lists... Done
/# apt-file search linux-vdso.so.1
```
2.   Could you check apache2 log under /var/log/apache2 to see if it download rootimg.cpio.gz or something wrong?  
The correct log process like:
```
xxxx - - [13/Dec/2018:00:30:55 -0500] "GET /tftpboot/xcat/xnba/nodes/c910f04x35v04 HTTP/1.1" 200 709 "-" "iPXE/1.0.3-130626"
xxxx - - [13/Dec/2018:00:30:55 -0500] "GET /tftpboot/xcat/osimage/ubuntu18.04.1-x86_64-netboot-compute/kernel HTTP/1.1" 200 8257483 "-" "iPXE/1.0.3-130626"
xxxx - - [13/Dec/2018:00:30:56 -0500] "GET /tftpboot/xcat/osimage/ubuntu18.04.1-x86_64-netboot-compute/initrd-stateless.gz HTTP/1.1" 200 163733285 "-" "iPXE/1.0.3-130626"
xxxx - - [13/Dec/2018:00:31:22 -0500] "GET //install/netboot/ubuntu18.04.1/x86_64/compute/rootimg.cpio.gz HTTP/1.1" 200 312224001 "-" "Wget/1.19.4 (linux-gnu)"
xxx - - [13/Dec/2018:00:31:35 -0500] "GET /install/postscripts/ HTTP/1.1" 200 28363 "-" "Wget/1.19.4 (
```

**Comment by jamesfreeman959** [comment]
Thank you both for your quick responses and confirming that the linux-vdso.so.1 errors can be ignored. I ran the boot process a second time, and it looks like it hangs after loading the initrd-stateless.gz image - there is no attempt to download rootimg.cpio.gz. I also note that there are no errors in /var/log/apache2/error.log

Here is the content from /var/log/apache2/access.log:

```
192.168.221.51 - - [13/Dec/2018:10:44:46 +0000] "GET /tftpboot/xcat/xnba/nodes/com1.uefi HTTP/1.1" 200 309 "-" "iPXE/1.0.3-130626"
192.168.221.51 - - [13/Dec/2018:10:44:46 +0000] "GET /tftpboot/xcat/elilo-x64.efi HTTP/1.1" 200 236326 "-" "iPXE/1.0.3-130626"
192.168.221.51 - - [13/Dec/2018:10:44:46 +0000] "GET /tftpboot/xcat/xnba/nodes/com1.elilo HTTP/1.1" 200 573 "-" "iPXE/1.0.3-130626"
192.168.221.51 - - [13/Dec/2018:10:44:46 +0000] "GET /tftpboot/xcat/osimage/ubuntu18.04.1-x86_64-netboot-compute/kernel HTTP/1.1" 200 8257483 "-" "iPXE/1.0.3-130626"
192.168.221.51 - - [13/Dec/2018:10:44:46 +0000] "GET /tftpboot/xcat/osimage/ubuntu18.04.1-x86_64-netboot-compute/initrd-stateless.gz HTTP/1.1" 200 163832808 "-" "iPXE/1.0.3-130626"
```

How can I help to debug this?

**Comment by immarvin** [comment]
hi @jamesfreeman959 , can you provide the console output during boot up?

**Comment by jamesfreeman959** [comment]
Hi there,

Thank you for this - I tried updating my xcat install from the snapshot builds, but the Debian.pm file referenced by https://github.com/xcat2/xcat-core/pull/5902/files did not change, so I simply patched it manually.

I then regenerated the image but booted the node, but it fails at the same place as before. The output right before the failure seems to be:

```
[    5.298334] bnx2x: QLogic 5771x/578xx 10/20-Gigabit Ethernet Driver bnx2x 1.712.30-0 (2014/02/10)
[    5.317686] e1000: Intel(R) PRO/1000 Network Driver - version 7.3.21-k8-NAPI
[    5.330033] e1000: Copyright (c) 1999-2006 Intel Corporation.
[    5.348411] e1000e: Intel(R) PRO/1000 Network Driver - 3.2.6-k
[    5.359593] e1000e: Copyright(c) 1999 - 2015 Intel Corporation.
[    5.377485] igb: Intel(R) Gigabit Ethernet Network Driver - version 5.4.0-k
[    5.389901] igb: Copyright (c) 2007-2014 Intel Corporation.
[    5.401092] usb 1-3.1: New USB device found, idVendor=0557, idProduct=7000
insmod: can't insert '/lib/mlx_en.ko': No such file or directory[    5.413539] usb 1-3.1: New USB device strings: Mfr=0, Product=0, SerialNumber=0

[    5.426910] hub 1-3.1:1.0: USB hub found
Creating device nodes with udev
/init: line 97: /sbin/udevd: not found
[    5.436550] hub 1-3.1:1.0: 4 ports detected
Couldn't find the proper booting device, switch to shell...
```

I note there is an insmod failure, but mlx_en.ko is a Mellanox module as I understand it and there is no Mellanox hardware on my system. Can this error be safely ignored? All network adapters look to have been successfully enumerated before this point (e1000e, bnx2x) so I believe the network is sound - however the failure still relates to not finding a proper boot device - presumably this is because the the root image has not been downloaded. 

Not sure how to proceed - I'm not clear where the failure is related to a driver load issue, or something else in the process.

Many thanks

James

**Comment by bybai** [comment]
Hi @jamesfreeman959 ,
 "can't insert '/lib/mlx_en.ko': No such file or directory" , this error can be safely ignored here since you do not have Mellanox hardware.

**Comment by immarvin** [comment]
hi @jamesfreeman959 , 

Please provide the /proc/cmdline content in the emergency shell.

the fix for the issue I mentioned has been included in the latest development build http://xcat.org/files/xcat/xcat-core/devel/Ubuntu/core-snap/20181228.0615-core-debs-snap.tar.bz2, if you are ok, you can take a try.

**Comment by jamesfreeman959** [comment]
Thanks for all your help everyone! I found the issue - my challenge was lack of familiarity with the xcat code, and confusion over the various error messages (again lack of familiarity as I don't know what can safely be ignored).

The mlx_en.ko error, and also the error regarding /sbin/udevd lead me to pull apart the initrd image. This in turn lead me to dissect the init script, and so I found my problem. In fact there were two - the first was resolved by pulling the latest snapshot as detailed previously. The next was actually rebuilding the image with the right network modules for my hardware. With these two fixes in place (the second entirely my fault!), I have a working stateless boot image.

One question - are there any plans to tidy up error messages? They make debugging problems difficult for someone new to xcat like me. For example /sbin/udevd hasn't existed since Ubuntu 14.04, so is there a way to remove it from the build of newer Ubuntu images, or do they all share the same script (and hence the line still exists?).

Thanks again!
